### PR TITLE
Fix "list_type_missing" API violations in meta/v1

### PR DESCRIPTION
**Pull request by thockin**
This assumes that any such field is atomic, except:
  * OwnerReferences: because it has a `+patchStrategy=merge`, but it probably needs a `+listMapKey=...` ?
  * Finalizers: because it hs a `+patchStrategy=merge`, but is a primitive type (string).

An alternative approach could be just to turn off the API warnings for these fields, but it felt more correct to declare the semantics.

If this flies I can do others.

@apelisse to tell me if I totally missed the mark on this assumption and to advise on the above.
@liggitt FYI
@deads2k For consult

/kind cleanup
```release-note
NONE
```

### CODE REVIEW

File: staging/src/k8s.io/apimachinery/pkg/runtime/types.go [hunk 1]
```diff
@@ -102,6 +102,7 @@ type RawExtension struct {
 	// Raw is the underlying serialization of this object.
 	//
 	// TODO: Determine how to detect ContentType and ContentEncoding of 'Raw' data.
+	// +listType=atomic
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/runtime/types.go:105]: these look ~fine for now, but we could also fix the linter to not complain about `[]byte` "lists". cc @apelisse

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 2]
```diff
@@ -236,6 +236,7 @@ type ObjectMeta struct {
 	// +optional
 	// +patchMergeKey=uid
 	// +patchStrategy=merge
+	// +listType=map
```
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:239]: I wondered about it.  In next push.

File: staging/src/k8s.io/apimachinery/pkg/runtime/types.go (hunk 1 above)
Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/runtime/types.go:105]: https://github.com/kubernetes/kube-openapi/pull/437

Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/runtime/types.go:105]: I will back these out from this PR, but as a new commit for readability

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 3]
```diff
@@ -268,6 +270,7 @@ type ObjectMeta struct {
 	// workflow used when modifying the object.
 	//
 	// +optional
+	// +listType=atomic
```
[the remark this answers is not in the document]
Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:273]: Would `+listType=None` make sense? (or something like that) ?

**Comment by thockin** [comment]
It would be nice if there was a doc which indicated the implicit defaults in each case.  E.g. if I set CSA "merge" on a `[]string` what is the correct SSA?

**Comment by thockin** [comment]
Now includes re-vendor of kube-openapi with no `[]byte` warnings.

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go [hunk 4]
```diff
@@ -1218,6 +1238,7 @@ type LabelSelector struct {
 	MatchLabels map[string]string `json:"matchLabels,omitempty" protobuf:"bytes,1,rep,name=matchLabels"`
 	// matchExpressions is a list of label selector requirements. The requirements are ANDed.
 	// +optional
+	// +listType=atomic
```
Comment (liggitt) [inline staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:1241]: @apelisse should the linter stop checking listType within a struct labeled as `+structType=atomic`?

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (liggitt) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:274]: in theory, these could still be included in an API of some type, so defining them as listType=atomic seems ok

File: staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go (hunk 4 above)
Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:1241]: I did not know `+structType=atomic` was a thing!  None of this seems to be documented.  I'll try removing that and see if it complains.

Comment (thockin) [reply staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/types.go:1241]: Nope, it still complains:

```
+API rule violation: list_type_missing,k8s.io/apimachinery/pkg/apis/meta/v1,LabelSelector,MatchExpressions
```

**Comment by liggitt** [comment]
changes lgtm, let me know when this is no longer WIP and you want a final check

**Comment by thockin** [comment]
I was going to keep chipping at it until I get them all, since we can't merge yet anyway

**Comment by thockin** [comment]
Opinion wanted: I'm now into things like "k8s.io/kube-controller-manager/config/v1alpha1/types.go" - these are not really APIs tha need "apply" logic, right?  Maybe the warning tool should not be considering them?

**Comment by liggitt** [comment]
> Opinion wanted: I'm now into things like "k8s.io/kube-controller-manager/config/v1alpha1/types.go" - these are not really APIs tha need "apply" logic, right? Maybe the warning tool should not be considering them? @liggitt @apelisse

There's bits in the warning tool (like field names matching serialized field names) that are useful even for config types. Agree the apply warnings don't really... apply... there.

**Comment by apelisse** [comment]
> There's bits in the warning tool (like field names matching serialized field names) that are useful even for config types. Agree the apply warnings don't really... apply... there.

;-)

What's the proper way to go then? A header to disable the warning? How do we know these types are "different"?

**Comment by liggitt** [comment]
(unit test flake is https://github.com/kubernetes/kubernetes/issues/121926)

**Comment by pastequo** [comment]
Should this PR be back-ported in 1.29 and others ? It might be useful for Customer Resources that includes native types into their definition, to have the full granularity for managedfields

https://kubernetes.slack.com/archives/C0123CNN8F3/p1705594643726679

**Comment by dprotaso** [comment]
Brought this up at sig-api-machinery this week and folks were ok with a cherry-picking this back ([notes](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix), [video](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix)) to a released version of K8s - in order to fix SSA

/cherrypick release-1.29

**Comment by liggitt** [comment]
> Brought this up at sig-api-machinery this week and folks were ok with a cherry-picking this back ([notes](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix), [video](https://docs.google.com/document/d/1x9RNaaysyO0gXHIr1y50QFbiL1x8OWnk2v3XnrdkT5Y/edit#heading=h.9p525fj57pix)) to a released version of K8s - in order to fix SSA

What specifically would this fix in SSA? the only two fields that actually changed anything effectively in this PR were ownerReferences and finalizers, right? neither of those were the ones mentioned in the linked slack thread

Backporting API behavior to a minor doesn't seem like a great idea to me. Letting the change take effect in the next release seems more reasonable.

**Comment by dprotaso** [comment]
> What specifically would this fix in SSA? the only two fields that actually changed anything effectively in this PR were ownerReferences and finalizers, right? neither of those were the ones mentioned in the linked slack thread

Various fields in the PodSpec now have the proper listType and listMapKey annotations - thus for those that embed this type into their own CRDs they'll have the proper OpenAPI schema when generating these CRD definitions.

**Comment by dprotaso** [comment]
> Backporting API behavior to a minor doesn't seem like a great idea to me. Letting the change take effect in the next release seems more reasonable.

Yeah that's fair - seems like the cherry-picker plugin isn't enabled anymore ;)

**Comment by liggitt** [comment]
> Various fields in the PodSpec now have the proper listType and listMapKey annotations - thus for those that embed this type into their own CRDs they'll have the proper OpenAPI schema when generating these CRD definitions.

Ah. Letting folks bump to k8s.io/api v0.30.0 when it's released would let them regen with annotations, right? I'd probably opt for that.

> seems like the cherry-picker plugin isn't enabled anymore ;)

it was never enabled for kubernetes/kubernetes, fyi

**Pull request merged**